### PR TITLE
gcc9: init at 9.1.0 (released today)

### PR DESCRIPTION
**Pull request by vcunat**
It's just a trivial copy of whole gcc8 stuff, only with changed version string and hash.

###### Motivation for this change

https://gcc.gnu.org/gcc-9/changes.html
I see nothing relevant to packaging in the changes.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] N/A Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] N/A Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of **some** binary files (usually in `./result/bin/`)
  [A small project](https://knot-resolver.cz) built fine with the resulting compiler.
- [ ] _I'll post below._ Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] N/A Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
/cc maintainer @Synthetica9.

### CODE REVIEW

**Comment by vcunat** [comment]
Closure size changes:
 - `.lib`: libstdc++ grew by almost half a megabyte (quite significant, but I can't see anything wrong)
 - `.out`: small increases in various places &ndash; static libstdc++, includes, binaries, ... +7 MiB of disk usage total
 - no changes in retained dependencies

File: pkgs/development/compilers/gcc/9/default.nix [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+  # Setting $CPATH and $LIBRARY_PATH to make sure both `gcc' and `xgcc' find the
+  # library headers and binaries, regarless of the language being compiled.
+  #
+  # Likewise, the LTO code doesn't find zlib.
+  #
+  # Cross-compiling, we need gcc not to read ./specs in order to build the g++
+  # compiler (after the specs for the cross-gcc are created). Having
+  # LIBRARY_PATH= makes gcc read the specs from ., and the build breaks.
+
+  CPATH = optionals (targetPlatform == hostPlatform) (makeSearchPathOutput "dev" "include" ([]
```
Comment (volth) [inline pkgs/development/compilers/gcc/9/default.nix]: ```optionals``` -> ```optional``` or ```optionalString```

```makeSearchPathOutput``` returns String, not List which is ```optionals``` seconds argument

File: pkgs/development/compilers/gcc/9/default.nix [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+    ++ optional (zlib != null) zlib
+  ));
+
+  LIBRARY_PATH = optionals (targetPlatform == hostPlatform) (makeLibraryPath (optional (zlib != null) zlib));
```
Comment (volth) [inline pkgs/development/compilers/gcc/9/default.nix]: here too

**Comment by vcunat** [comment]
We can fix these, but in all the copies.  I see no reason to differ in this respect.

**Comment by dtzWill** [comment]
No longer need musl patch, didn't check history but doesn't apply and matches my memory of things: f6344621a00a12bc7f97bbc515736e0623cac973 

I can submit separately if you'd prefer, but adding a `gcc9Stdenv` attribute and entry in the cc-wrapper tests: 3ac768c620c2c602a02cb6fab3b23d69e5818d58

**Comment by dtzWill** [comment]
Otherwise LGTM.  Didn't vet all the patch / configuration options (powerpc musl patch?), but at least the x86_64 glibc/musl seem good :+1:.

**Comment by vcunat** [comment]
OK, your two commits (or equivalent) were pushed post-merge.  Support of gcc9 for less usual **languages**:
 - fortran is waiting for feedback: https://github.com/NixOS/nixpkgs/issues/61130
 - go would suffer from cycle between .out and .lib (same as some older versions IIRC), and there seems not enough demand
 - D is now supported upstream, but I've done nothing around that
 - Ada was abandoned by us long ago &ndash; due to difficult bootstrapping and practically no demand

**Pull request merged**